File: kotlin/app/src/test/kotlin/org/example/ConfirmCheckoutTest.kt
package org.example

import org.junit.jupiter.api.Test
import java.util.Hashtable
import java.util.UUID

class ConfirmCheckoutTest {
    @Test
    fun dummy() {
        val purchaseOrderRepository = DummyPurchaseOrderRepository()
        ConfirmCheckout(
            uuidGenerator = UUIDGenerator(),
            purchaseOrderRepository = purchaseOrderRepository,
            pricingService = TODO(),
            paymentService = TODO(),
            invoicingService = TODO(),
            shippingService = TODO(),
            stockService = TODO()
        )
    }

    class DummyPurchaseOrderRepository : PurchaseOrderRepository {
        override fun save(purchase: Purchase): Purchase {
            TODO()
        }
    }

    @Test
    fun fake() {
        val purchaseOrderRepository = FakePurchaseOrderRepository()
        ConfirmCheckout(
            uuidGenerator = UUIDGenerator(),
            purchaseOrderRepository = purchaseOrderRepository,
            pricingService = TODO(),
            paymentService = TODO(),
            invoicingService = TODO(),
            shippingService = TODO(),
            stockService = TODO()
        )
    }

    class FakePurchaseOrderRepository: PurchaseOrderRepository {
        private val hashtable = Hashtable<UUID, Purchase>()
        override fun save(purchase: Purchase): Purchase {
            hashtable[purchase.id] = purchase
            return purchase
        }
    }

    @Test
    fun stub() {
        val purchaseOrderRepository = mock()
        ConfirmCheckout(
            uuidGenerator = UUIDGenerator(),
            purchaseOrderRepository = purchaseOrderRepository,
            pricingService = TODO(),
            paymentService = TODO(),
            invoicingService = TODO(),
            shippingService = TODO(),
            stockService = TODO()
        )
    }

    class FakePurchaseOrderRepository: PurchaseOrderRepository {
        private val hashtable = Hashtable<UUID, Purchase>()
        override fun save(purchase: Purchase): Purchase {
            hashtable[purchase.id] = purchase
            return purchase
        }
    }
}

File: scala/src/test/scala/workshop/ConfirmCheckoutTest.scala
package workshop

import com.workshop.{ConfirmCheckout, StockService, UUIDGenerator}
import dummies.{DummyInvoicingService, DummyPaymentService, DummyPricingService, DummyShippingService, DummyStockService}
import fakes.InMemPurchaseOrderRepository
import org.scalatest.funspec.AnyFunSpec
import org.scalamock.scalatest.MockFactory

import scala.util.Try

class ConfirmCheckoutTest extends AnyFunSpec with MockFactory {
  describe("ConfirmCheckout") {
    val uuidGenerator = new UUIDGenerator()

    // Dummies
    val stockService = new DummyStockService()
    val shippingService = new DummyShippingService()
    val invoicingService = new DummyInvoicingService()
    val paymentService = new DummyPaymentService()
    val pricingService = new DummyPricingService()

    // Fake
    val inMemoryOrderRepository = new InMemPurchaseOrderRepository()

    it("should confirm a checkout successfully") {
      // Dummies: StockService, PaymentService, PricingService
      // Mocks: ShippingService, InvoicingService

      val shippingService = mock[DummyShippingService]
      val invoicingService = mock[DummyInvoicingService]

      (shippingService.ship _).expects(*, *).returning(*).once()
      (invoicingService.issue _).expects(*).returning(*).once()

      val confirmCheckout = new ConfirmCheckout(
        uuidGenerator,
        inMemoryOrderRepository,
        pricingService,
        paymentService,
        invoicingService,
        shippingService,
        stockService
      )

      val executeResult = Try {
        confirmCheckout.execute(
          List(("product1", 1)),
          "1234567890123456",
          2024,
          12,
          "123",
          "John Doe",
          "123 Main St",
          "john.doe@acme.org",
          "customer1")
      }

      assert(executeResult.isSuccess)
    }

    it("should fail to confirm a checkout due to insufficient stock") {
      val mockStockService = stub[StockService]

      // Stub
      (mockStockService.checkWhetherEnoughStock _).when(*).returns(false)

      val confirmCheckout = new ConfirmCheckout(
        uuidGenerator,
        inMemoryOrderRepository,
        pricingService,
        paymentService,
        invoicingService,
        shippingService,
        mockStockService
      )

      val executeResult = Try {
        confirmCheckout.execute(
          List(("product1", 1)),
          "1234567890123456",
          2024,
          12,
          "123",
          "John Doe",
          "123 Main St",
          "john.doe@acme.org",
          "customer1")
      }

      assert(executeResult.isFailure)
      assert(executeResult.failed.toOption.forall(_.isInstanceOf[RuntimeException]))
    }
  }
}
